HPCC-35296 Improve AWS S3 implementation

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-35296 Improve AWS S3 implementation

jackdelv commented on common/remote/hooks/s3/s3file.cpp:
```diff
@@ -95,293 +70,735 @@ static void ensureAWSInitialized()
+            auto body = std::make_shared<std::stringstream>(std::ios::in | std::ios::out | std::ios::binary);
+            body->write(reinterpret_cast<const char*>(data), len);
+            if (!body->good() || body->tellp() != static_cast<std::streampos>(len))
+                throw makeStringException(-1, "Failed to prepare multipart upload buffer");
```
I don't think it needs to be seeked to the beginning because the files uploaded are not empty. Also, it looks like it is casting it to a buffer which then passes the pointer to the data.

jackdelv commented on common/remote/hooks/s3/s3file.hpp:
```diff
@@ -27,13 +28,54 @@
  */
 
+// Forward declarations
+interface IPropertyTree;
```
Will fix.

jackdelv commented on common/remote/hooks/s3/s3file.cpp:
```diff
@@ -95,293 +70,735 @@ static void ensureAWSInitialized()
     {
-        throwUnexpected();
+        return (strcmp(config.region.str(), currentConfig.region.str()) != 0) ||
+               (strcmp(config.endpoint.str(), currentConfig.endpoint.str()) != 0) ||
```
Changed to use strsame instead.

jackdelv commented on common/remote/hooks/s3/s3file.hpp:
```diff
@@ -27,13 +28,54 @@
+// Configuration structure for S3 operations
+struct S3Config
+{
+    IPropertyTree* config = nullptr;
```
Removed config member

ghalliday reviewed (changes requested):
@jackdelv I haven't reviewed the code, except for looking at the way configuration is supported.   Please IM me to arrange to talk if that doesn't make sense, or want to talk through my comments.

jackdelv commented:
@ghalliday I have added changes that allow for the client connection to be created using credentials and settings from the storage plane. I added a similar scheme to Azure where a Storage plane can have multiple buckets with their own access keys.

During testing, I ran into a few issues. When running on HThor, it use the AWS API and instead wrote a local file. Also, when running on Thor, I kept getting an error because numDevices was set to zero, but the stripe number was not included in the logical file name.

Changing dadesc.cpp:1421 from `unsigned numDevices = cluster ? cluster->queryPartDiskMapping().numStripedDevices : plane->numDevices();` to

```
// For cloud storage planes (S3, Azure, etc.), prefer the plane's numDevices over the cluster's setting
unsigned numDevices = plane->numDevices();
if (cluster)
{
    unsigned clusterDevices = cluster->queryPartDiskMapping().numStripedDevices;
    if (clusterDevices > numDevices)
        numDevices = clusterDevices;
}
```
allowed the striped number to be included in the filename.

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
Again, avoid user logging - e.g. it will appear in the output of eclcc.

jackdelv replied:
Fixed

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
Almost certainly too late to log the problem.  Did you see this happening?

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
equality is not implemented for StringAttr - so need to call strsame().  Better to use std::string.

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
new code should use queryStoragePlaneByName(), but the interface may need extending to provide the information that is needed.  Better revisited as a separate PR.

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
This test looks redundant/inefficient

jackdelv replied:
Removed redundant check

ghalliday commented on common/remote/hooks/s3/s3file.cpp:
This will clone the data.  Is there an alternative api that will avoid that?

ghalliday reviewed (changes requested):
@jackdelv that looks good - much closer to ready.
A few comments about style and structure and no need to buffer.

jackdelv commented on common/remote/hooks/s3/s3file.cpp:
Fixed

jackdelv commented:
Yes, every part except for the last one needs to be 5MB when writing to S3.

ghalliday commented:
I think it would be reasonable to throw an error if 
getBlockedFileIOSize(planeName, 0)
is < 5MB.
That should ensure good values are used for read and write.

ghalliday reviewed (changes requested):
@jackdelv there is a windows build issue that needs fixing

jackdelv commented:
@ghalliday Do you know why I'm all of a sudden getting way more errors? I was pretty certain my change fixed the problem, but these are all new errors.
